File: Two Pointers/392_Is-Subsequence/392_Is-Subsequence.py
"""
 * Title:
 *     392. Is Subsequence
 *     392. 判断子序列
 * Address: 
 *     https://leetcode-cn.com/problems/is-subsequence/
"""

# 方法三：官方解法
# https://leetcode-cn.com/problems/is-subsequence/solution/pan-duan-zi-xu-lie-by-leetcode-solution/
class Solution_3:
    def isSubsequence(self, s: str, t: str) -> bool:
        lenS, lenT = len(s), len(t)
        i = j = 0

        while i < lenS and j < lenT:
            if s[i] == t[j]:
                i += 1
            j += 1
        
        return i == lenS

# 方法二：逻辑更清晰
class Solution_2:
    def isSubsequence(self, s: str, t: str) -> bool:
        if len(s) == 0:
            return True
        elif len(t) == 0:  # 此时 len(s) != 0
            return False
        
        i = 0
        for j in range(len(t)):
            if s[i] == t[j]:
                i += 1
                if i == len(s):
                    return True
        
        return False

# 方法一：
class Solution:
    def isSubsequence(self, s: str, t: str) -> bool:
        if len(s) == 0:
            return True
        elif len(t) == 0:
            return False
        
        lenS = len(s)
        lenT = len(t)
        j = 0
        for i in range(lenS):
            while j < lenT:
                if s[i] == t[j]:
                    j += 1
                    break
                else:
                    j += 1
            
            if j == lenT and (i != lenS - 1 or s[-1] != t[-1]):
                return False
        
        return True

File: Two Pointers/392_Is-Subsequence/392_Is-Subsequence.cpp
/*******************************************************************************
 * Title:
 *     392. Is Subsequence
 *     392. 判断子序列
 * Address: 
 *     https://leetcode-cn.com/problems/is-subsequence/
 ******************************************************************************/

#include <iostream>
#include <vector>
using namespace std;

// 方法三：官方解法
// https://leetcode-cn.com/problems/is-subsequence/solution/pan-duan-zi-xu-lie-by-leetcode-solution/
class Solution_3 {
public:
    bool isSubsequence(string s, string t) {
        int lenS = s.size(), lenT = t.size();
        int i = 0, j = 0;

        while ( i < lenS && j < lenT ) {
            if ( s[i] == t[j] )
                ++i;
            ++j;
        }

        return i == lenS;
    }
};

// 方法二：逻辑更清晰
class Solution_2 {
public:
    bool isSubsequence(string s, string t) {
        if ( s.empty() )
            return true;
        else if ( t.empty() )
            return false;
        
        int i = 0, j = 0;

        for ( int j = 0; j < (int)t.size(); ++j ) {
            if ( s[i] == t[j] ) {
                ++i;
                if ( i == (int)s.size() )
                    return true;
            }
        }

        return false;
    }
};


// 方法一：
class Solution {
public:
    bool isSubsequence(string s, string t) {
        if ( s.empty() )
            return true;
        else if ( t.empty() )  // 此时 s 
            return false;

        int lenS = s.size(), lenT = t.size();
        int i = 0, j = 0;

        for ( i = 0; i < lenS; ++i ) {
            while ( j < lenT ) {
                if ( s[i] == t[j++] ) {  // 进入 if 语句后，会执行 j++
                    break;
                }
            }

            // j == lenT : 此时遍历到 t 末端
            if ( j == lenT && ( i != lenS - 1 || s.back() != t.back() ) ) {
                return false;
            }
        }

        return true;
    }
};



int main()
{
    

    return 0;
}